src/lib/rules/apple-touch-icons/apple-touch-icons.ts

```diff
@@ -0,0 +1,268 @@
+                //
+                //  but that donesn't work in practice.
+
+                const relValue = element.getAttribute('rel');
+
+                if (relValue === null) {
+                    return false;
+                }
+
+                const relValues = normalizeString(element.getAttribute('rel')).split(' ');
+
+                return relValues.includes('apple-touch-icon') || relValues.includes('apple-touch-icon-precomposed');
+            });
+        };
+
+        const checkImage = async (appleTouchIcon: IAsyncHTMLElement, resource: string) => {
+            const appleTouchIconHref = normalizeString(appleTouchIcon.getAttribute('href'));
+
+            // Check if `href` doesn't exist, or it has the
+            // value of empty string.
+
+            if (!appleTouchIconHref) {
+                await context.report(resource, appleTouchIcon, `'apple-touch-icon' should have non-empty 'href' attribute`);
+
+                return;
+            }
+
+            // - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+            // The following checks don't make sense for non-HTTP(S).
+
+            if (!isRegularProtocol(resource)) {
+                return;
+            }
+
+            // - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+            let appleTouchIconURL = '';
+
+            // If `href` exists and is not an empty string, try
+            // to figure out the full URL of the `apple-touch-icon`.
+
+            if (url.parse(appleTouchIconHref).protocol) {
+                appleTouchIconURL = appleTouchIconHref;
+            } else {
+                appleTouchIconURL = url.resolve(resource, appleTouchIconHref);
+            }
+
+            // - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+            let networkData: INetworkData;
+
+            // Try to see if the `apple-touch-icon` file actually
+            // exists and is accesible.
+
+            try {
+                networkData = await context.fetchContent(appleTouchIconURL);
+            } catch (e) {
+                debug(`Failed to fetch the ${appleTouchIconHref} file`);
+                await context.report(resource, appleTouchIcon, `'${appleTouchIconHref}' file request failed`);
+
+                return;
+            }
+
+            const response = networkData.response;
+
+            if (response.statusCode !== 200) {
+                await context.report(resource, appleTouchIcon, `'${appleTouchIconHref}' could not be fetched (status code: ${response.statusCode})`);
+
+                return;
+            }
+
+            // - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+
+            // Note: Async version of `image-size` doesn't work if the
+            // input is a Buffer.
+            //
+            // https://github.com/image-size/image-size/tree/4c527ba608d742fbb29f6d9b3c77b831b069cbb2#asynchronous
+
+            const image = getImageData(response.body.rawContent);
```

Comment:
`image-size` exposes an [async API](https://github.com/image-size/image-size#asynchronous) that can be promisified. I think we should use that:

```js
import { promisify } from 'util';
import * as getImageData from 'image-size';

const getImageDataAsync = promisify(getImageData);

...

const image = await getImageData(response.body.rawContent);
```

Reply:
@austinkelleher See [my comment](https://github.com/sonarwhal/sonar/pull/488/files/9b9900abee8adc44d6956f876b0e7a0014c5f516#diff-046ee208e76bb5dffb66c554f2aac000R118).